Fix trees

nyee opened the pull request:
Some changes related to my attempts to fix the trees as related to RMG-database issue

rwest commented:
I made some suggestions. Once you've checked them, @nyee, you should be able to
`git rebase -i --autosquash master fixTrees`  and it will put them into place as fixup commits and merge them with yours.

nyee commented:
Thanks for the suggestions, Richard. I have merged them in and forced pushed the updated version to this branch.

nyee commented:
This latest test passes for the RMG-database branch [fixTrees](https://github.com/nyee/RMG-database/tree/fixTrees).  The same test should be implemented on kinetics, solvation, transport, and statmech also, but I didn't want to put it in yet because no RMG-database commit will pass those tests.

In terms of debugging those other databases: It is easier to use this script to automatically rearrange the trees:

[fixSiblings.txt](https://github.com/ReactionMechanismGenerator/RMG-Py/files/298335/fixSiblings.txt)

The above script should **not** be trusted used blindly. For example, it might be that groups are not [explicit enough](https://github.com/nyee/RMG-database/commit/71ea6851397cfb808a41ea7e6812a1e50ca1b7b4), rather than the tree is set up wrong.

KEHANG commented on rmgpy/data/base.py:
```diff
@@ -875,10 +875,10 @@ def matchNodeToChild(self, parentNode, childNode):
         #family line is consistent, which is done in the databaseTest unitTest
         elif isinstance(parentNode.item, Group) and isinstance(childNode.item, LogicOr):
-            return True
+            return childNode.parent is parentNode
```
in order to spot possible suspicious parent/child, for this if situation `isinstance(parentNode.item, Group) and isinstance(childNode.item, LogicOr)`, we  might also want to check the items in the `LogicOr` are children of the `parentNode`

nyee replied:
As said in the comment above, we actually do this in [databaseTest.py](https://github.com/ReactionMechanismGenerator/RMG-Py/blob/059b895b67370b2daa5699c3b5c84ffc87a36add/testing/databaseTest.py#L245-L253
l). 

In my opinion, the current implementation, unless we want to start using the matchNodeToChild for things other than database unittests.

KEHANG commented on rmgpy/data/thermo.py:
```diff
@@ -423,6 +423,53 @@ def processOldLibraryEntry(self, data):
+
+        #extend parent's children attribute with new children
+        parentR.children.extend(entryToRemove.children)
+
```
please add an unittest for this helpful method

KEHANG commented on testing/databaseTest.py:
```diff
@@ -259,6 +277,26 @@ def kinetics_checkChildParentRelationships(self, family_name):
                     nose.tools.assert_true(family.matchNodeToChild(ancestorNode, childNode),
                                     "In {family} family, group {ancestor} is not a proper ancestor of its child {child}.".format(family=family_name, ancestor=ancestorNode, child=nodeName))
 
+    def kinetics__checkSiblingsForParents(self, family_name):
```
remove one underscore in the method name?

nyee replied:
Did some rebasing and this is now fixed in commit 1bba6c

KEHANG commented on rmgpy/data/kinetics/common.py:
```diff
@@ -83,40 +83,44 @@ def sortEfficiencies(efficiencies0):
+
+
     if isinstance(entry.item, Reaction):
-        if entry.label != str(entry.item):
```
seems this `saveEntry` method is used for many purposes, can we also add some comments in the method to clarify which case (e.g., save to `groups.py`, `rules.py`, `reactions.py`) is for which if statement?

nyee replied:
Fixed up in commit fdb29

nyee commented:
Added unittest for removeGroup with latest commit. Discovered some bugs along the way and fixed them in the fixup! commit for the removeGroup method.

Unfortunately, I can't run make test on my computer, and I'm not sure I used the correct path to the test_data, so I need travis to tell me whether the unittest passes.

KEHANG commented on rmgpy/data/thermo.py:
```diff
@@ -421,6 +421,66 @@ def processOldLibraryEntry(self, data):
+                parentR.referenceType = groupToRemove.referenceType
+
+        #look for other pointers that point toward entry
+        for entryName, entry in self.entries.iteritems():
```
I think this `for` should be combined with the `if` statement starting at line 457 in order to make it more concise.
